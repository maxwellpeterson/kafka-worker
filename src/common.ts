export interface Env {
  // The hostname of this worker
  HOSTNAME: string;
  CLUSTER: DurableObjectNamespace;
  // Example binding to KV. Learn more at https://developers.cloudflare.com/workers/runtime-apis/kv/
  // MY_KV_NAMESPACE: KVNamespace;
  //
  // Example binding to R2. Learn more at https://developers.cloudflare.com/workers/runtime-apis/r2/
  // MY_BUCKET: R2Bucket;
}

export type ValueOf<T> = T[keyof T];
