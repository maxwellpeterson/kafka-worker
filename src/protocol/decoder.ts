import {
  Acks,
  ErrorCode,
  Int16,
  Int32,
  Int64,
  KafkaArray,
  MessageSet,
  NullableString,
  int16Size,
  int32Size,
  int64Size,
  validAcks,
  validErrorCode,
} from "src/protocol/common";

// This implementation borrows heavily from the kafkajs Node library:
// https://github.com/tulios/kafkajs/blob/master/src/protocol/decoder.js

export type EnumPredicate<T extends Int16> = (value: Int16) => value is T;

export class Decoder {
  private readonly view: DataView;
  private offset: number;

  constructor(buffer: ArrayBuffer) {
    this.view = new DataView(buffer);
    this.offset = 0;
  }

  readInt16(): Int16 {
    const value = this.view.getInt16(this.offset);
    this.offset += int16Size;
    return value;
  }

  readEnum<T extends Int16>(predicate: EnumPredicate<T>): T {
    const value = this.readInt16();
    if (!predicate(value)) {
      throw new Error(`Invalid enum value: ${value}`);
    }
    return value;
  }

  readErrorCode(): ErrorCode {
    return this.readEnum(validErrorCode);
  }

  readAcks(): Acks {
    return this.readEnum(validAcks);
  }

  readInt32(): Int32 {
    const value = this.view.getInt32(this.offset);
    this.offset += int32Size;
    return value;
  }

  readInt64(): Int64 {
    const value = this.view.getBigInt64(this.offset);
    this.offset += int64Size;
    return value;
  }

  private readStringSize(size: number): string {
    const stringBuffer = this.view.buffer.slice(
      this.offset,
      this.offset + size
    );
    const value = new TextDecoder().decode(stringBuffer);
    this.offset += size;
    return value;
  }

  readString(): string {
    const size = this.readInt16();
    if (size === -1) {
      throw new Error("Unexpected null string!");
    }
    return this.readStringSize(size);
  }

  readNullableString(): NullableString {
    const size = this.readInt16();
    if (size === -1) {
      return null;
    }
    return this.readStringSize(size);
  }

  readKafkaArray<T>(readElement: (index: number) => T): KafkaArray<T> {
    const size = this.readInt32();
    if (size === -1) {
      return null;
    }
    const values = new Array<T>(size);
    for (let i = 0; i < size; i++) {
      values[i] = readElement(i);
    }
    return values;
  }

  // Convenience method for converting null array to empty array in cases where
  // there is no semantic difference between them
  readArray<T>(readElement: (index: number) => T): T[] {
    return this.readKafkaArray(readElement) ?? [];
  }

  readInt32Array(): Int32[] {
    return this.readArray(() => this.readInt32());
  }

  readInt64Array(): Int64[] {
    return this.readArray(() => this.readInt64());
  }

  readStringArray(): string[] {
    return this.readArray(() => this.readString());
  }

  private readSlice(size: Int32): Uint8Array {
    const slice = new Uint8Array(this.view.buffer, this.offset, size);
    this.offset += size;
    return slice;
  }

  readMessageSet(): MessageSet {
    const size = this.readInt32();
    return this.readSlice(size);
  }
}
