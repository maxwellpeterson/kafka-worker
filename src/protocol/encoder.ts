import {
  Int16,
  Int32,
  Int64,
  Int8,
  KafkaArray,
  MessageSet,
  NullableString,
  int16Size,
  int32Size,
  int64Size,
  int8Size,
} from "src/protocol/common";

// This implementation borrows heavily from the kafkajs Node library:
// https://github.com/tulios/kafkajs/blob/master/src/protocol/encoder.js

export class Encoder {
  protected view: DataView;
  protected offset: number;

  constructor(initialBufferSize = 64) {
    const buffer = new ArrayBuffer(initialBufferSize);
    this.view = new DataView(buffer);
    this.offset = 0;
  }

  private checkCapacity(size: number) {
    const nextLength = this.offset + size;
    if (nextLength > this.view.byteLength) {
      // If we run out of space, reallocate the backing array
      const newCapacity = 2 * nextLength;
      const newBuffer = new ArrayBuffer(newCapacity);
      new Uint8Array(newBuffer).set(new Uint8Array(this.view.buffer));
      this.view = new DataView(newBuffer);
    }
  }

  writeInt8(value: Int8): this {
    this.checkCapacity(int8Size);
    this.view.setInt8(this.offset, value);
    this.offset += int8Size;
    return this;
  }

  writeInt16(value: Int16): this {
    this.checkCapacity(int16Size);
    this.view.setInt16(this.offset, value);
    this.offset += int16Size;
    return this;
  }

  // Convenience method that makes sure "enum" values are encoded as Int16
  writeEnum<T extends Int16>(value: T): this {
    return this.writeInt16(value);
  }

  writeInt32(value: Int32): this {
    this.checkCapacity(int32Size);
    this.view.setInt32(this.offset, value);
    this.offset += int32Size;
    return this;
  }

  writeInt64(value: Int64): this {
    this.checkCapacity(int64Size);
    this.view.setBigInt64(this.offset, value);
    this.offset += int64Size;
    return this;
  }

  writeString(value: string): this {
    const bytes = new TextEncoder().encode(value);
    this.writeInt16(bytes.length);
    this.checkCapacity(bytes.length);
    new Uint8Array(this.view.buffer).set(bytes, this.offset);
    this.offset += bytes.length;
    return this;
  }

  writeNullableString(value: NullableString): this {
    if (value === null) {
      this.writeInt16(-1);
      return this;
    }
    return this.writeString(value);
  }

  writeArray<T>(values: KafkaArray<T>, writeElement: (value: T) => void): this {
    if (values === null) {
      this.writeInt32(-1);
      return this;
    }
    this.writeInt32(values.length);
    values.forEach(writeElement);
    return this;
  }

  writeInt32Array(values: Int32[]): this {
    return this.writeArray(values, (value) => this.writeInt32(value));
  }

  writeInt64Array(values: Int64[]): this {
    return this.writeArray(values, (value) => this.writeInt64(value));
  }

  writeStringArray(values: KafkaArray<string>): this {
    return this.writeArray(values, (value) => this.writeString(value));
  }

  writeSlice(slice: Uint8Array): this {
    this.checkCapacity(slice.length);
    new Uint8Array(this.view.buffer).set(slice, this.offset);
    this.offset += slice.length;
    return this;
  }

  writeBytes(bytes: Uint8Array): this {
    this.writeInt32(bytes.length);
    return this.writeSlice(bytes);
  }

  writeMessageSet(messageSet: MessageSet): this {
    return this.writeBytes(messageSet);
  }

  buffer(): ArrayBuffer {
    if (this.view.buffer.byteLength === this.offset) {
      // Avoid an allocation/copy if the underlying buffer does not need to be
      // resized (probably because initialBufferSize was set optimally)
      return this.view.buffer;
    }
    return this.view.buffer.slice(0, this.offset);
  }

  slice(): Uint8Array {
    return new Uint8Array(this.view.buffer, 0, this.offset);
  }
}
