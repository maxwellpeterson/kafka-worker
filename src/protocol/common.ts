import { ValueOf } from "src/common";

// These don't add any type safety, but help with labeling
export type Int16 = number;
export type Int32 = number;
export type Int64 = bigint;

export const int16Size = 2;
export const int32Size = 4;
export const int64Size = 8;

export type KafkaArray<T> = T[] | null;

export interface RequestMetadata {
  apiVersion: Int16;
  correlationId: Int32;
  clientId: string;
}
export const generateEnumPredicate =
  <E extends Record<string, V>, V>(enumObj: E) =>
  (value: V): value is ValueOf<E> =>
    Object.values<V>(enumObj).includes(value);

// ApiKey is an Int16
export const ApiKey = {
  Produce: 0,
  // Fetch: 1,
  // ListOffsets: 2,
  Metadata: 3,
} as const;
export type ApiKey = ValueOf<typeof ApiKey>;
export const validApiKey = generateEnumPredicate(ApiKey);

// ErrorCode is an Int16
export const ErrorCode = {
  None: 0,
  UnknownTopicOrPartition: 3,
} as const;
export type ErrorCode = ValueOf<typeof ErrorCode>;
export const validErrorCode = generateEnumPredicate(ErrorCode);

// Acks is an Int16
export const Acks = {
  None: 0,
  Leader: 1,
  FullISR: -1,
} as const;
export type Acks = ValueOf<typeof Acks>;
export const validAcks = generateEnumPredicate(Acks);
