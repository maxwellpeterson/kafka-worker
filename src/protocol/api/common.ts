import { Env } from "src/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

export type EntrypointFn<Output> = (
  env: Env,
  version: number,
  decoder: Decoder,
  encoder: Encoder
) => Promise<Output>;

export type CheckVersionFn = (version: number) => void;
export type DecodeFn<Req> = (decoder: Decoder) => Req;
export type HandleFn<Req, Res> = (env: Env, request: Req) => Promise<Res>;
export type EncodeFn<Res, Output> = (encoder: Encoder, response: Res) => Output;

export const generateEntrypointFn =
  <Req, Res, Output>(
    apiName: string,
    checkVersion: CheckVersionFn,
    decode: DecodeFn<Req>,
    handle: HandleFn<Req, Res>,
    encode: EncodeFn<Res, Output>
  ): EntrypointFn<Output> =>
  async (
    env: Env,
    version: number,
    decoder: Decoder,
    encoder: Encoder
  ): Promise<Output> => {
    checkVersion(version);

    const request = decode(decoder);
    console.log(
      `Received ${apiName} request: ${JSON.stringify(request, null, 2)}`
    );

    const response = await handle(env, request);
    console.log(
      `Sending ${apiName} response: ${JSON.stringify(response, null, 2)}`
    );

    return encode(encoder, response);
  };

export const generateCheckVersionFn =
  (apiName: string, expected: number): CheckVersionFn =>
  (version: number) => {
    if (version !== expected) {
      throw new Error(
        `Unsupported version of ${apiName} api: expected ${expected} but got ${version}`
      );
    }
  };
