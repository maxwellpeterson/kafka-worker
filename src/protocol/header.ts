import { Int16, Int32, NullableString } from "src/protocol/common";
import { Decoder, EnumPredicate } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

// Request Header (Version: 1) => api_key api_version correlation_id client_id
//   api_key => INT16
//   api_version => INT16
//   correlation_id => INT32
//   client_id => NULLABLE_STRING
//
// https://kafka.apache.org/protocol.html#protocol_messages
//
// This header format is also used by internal requests.

export interface RequestHeader<T> {
  apiKey: T;
  apiVersion: Int16;
  correlationId: Int32;
  clientId: NullableString;
}

export const encodeRequestHeader = <T extends Int16>(
  encoder: Encoder,
  header: RequestHeader<T>
): Encoder => {
  return encoder
    .writeEnum(header.apiKey)
    .writeInt16(header.apiVersion)
    .writeInt32(header.correlationId)
    .writeNullableString(header.clientId);
};

export const decodeRequestHeader = <T extends Int16>(
  decoder: Decoder,
  predicate: EnumPredicate<T>
): RequestHeader<T> => {
  return {
    apiKey: decoder.readEnum(predicate),
    apiVersion: decoder.readInt16(),
    correlationId: decoder.readInt32(),
    clientId: decoder.readNullableString(),
  };
};

export type RequestMetadata = Omit<RequestHeader<unknown>, "apiKey">;
