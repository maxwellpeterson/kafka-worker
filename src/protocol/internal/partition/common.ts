import { ValueOf } from "src/common";
import {
  ApiKey,
  Int16,
  Int32,
  generateEnumPredicate,
  int32Size,
} from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

// Partition Request (Version: 0) => header request_message
//   header => api_key api_version correlation_id client_id
//     api_key => INT16
//     api_version => INT16
//     correlation_id => INT32
//     client_id => STRING
//   request_message => PARTITION_PRODUCE_REQUEST

export interface PartitionRequestHeader {
  apiKey: PartitionApiKey;
  apiVersion: Int16;
  correlationId: Int32;
  clientId: string;
}

export const encodePartitionRequestHeader = (
  encoder: Encoder,
  header: PartitionRequestHeader
) => {
  encoder.writeEnum(header.apiKey);
  encoder.writeInt16(header.apiVersion);
  encoder.writeInt32(header.correlationId);
  encoder.writeString(header.clientId);
};

export const decodePartitionRequestHeader = (
  decoder: Decoder
): PartitionRequestHeader => {
  return {
    apiKey: decoder.readEnum(validPartitionApiKey),
    apiVersion: decoder.readInt16(),
    correlationId: decoder.readInt32(),
    clientId: decoder.readString(),
  };
};

// PartitionApiKey is an Int16
export const PartitionApiKey = {
  Produce: ApiKey.Produce,
} as const;
export type PartitionApiKey = ValueOf<typeof PartitionApiKey>;
const validPartitionApiKey = generateEnumPredicate(PartitionApiKey);

// Partition Response (Version: 0) => correlation_id response_message
//   correlation_id => INT32
//   response_message => PARTITION_PRODUCE_RESPONSE

export class PartitionResponseEncoder extends Encoder {
  constructor(correlationId: Int32, initialBufferSize = 64) {
    super(initialBufferSize + int32Size);
    this.writeInt32(correlationId);
  }
}
