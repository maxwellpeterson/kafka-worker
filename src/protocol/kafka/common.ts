import { Int32, int32Size } from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

export class KafkaResponseEncoder extends Encoder {
  constructor(correlationId: Int32, initialBufferSize = 64) {
    super(initialBufferSize + 2 * int32Size);
    // Reserve space for size header at front of buffer
    this.offset += int32Size;
    this.writeInt32(correlationId);
  }

  // Overrides parent method, adds size header to buffer
  buffer(): ArrayBuffer {
    // Write size header to reserved space at front of buffer
    this.view.setInt32(0, this.offset - int32Size);
    return super.buffer();
  }
}

export class KafkaRequestDecoder extends Decoder {
  constructor(buffer: ArrayBuffer) {
    super(buffer);

    const expectedSize = this.readInt32();
    const actualSize = buffer.byteLength - int32Size;

    if (expectedSize !== actualSize) {
      throw new Error(
        `Message length does not match size header: expected ${expectedSize} but got ${actualSize}`
      );
    }
  }
}
