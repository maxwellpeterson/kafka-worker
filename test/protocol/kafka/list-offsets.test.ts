import { ErrorCode } from "src/protocol/common";
import {
  KafkaListOffsetsRequest,
  KafkaListOffsetsResponse,
  decodeKafkaListOffsetsRequest,
  decodeKafkaListOffsetsResponse,
  encodeKafkaListOffsetsRequest,
  encodeKafkaListOffsetsResponse,
} from "src/protocol/kafka/list-offsets";
import { testEncodeDecodeSnapshot } from "test/protocol/common";

describe("KafkaListOffsetsRequest", () => {
  type TestCase = [string, KafkaListOffsetsRequest];
  const cases: TestCase[] = [
    [
      "latest offset",
      {
        replicaId: 333,
        topics: [
          {
            name: "topic-one",
            partitions: [
              {
                index: 0,
                timestamp: BigInt(-1),
                maxNumOffsets: 1,
              },
            ],
          },
        ],
      },
    ],
    [
      "earliest offset",
      {
        replicaId: 333,
        topics: [
          {
            name: "topic-one",
            partitions: [
              {
                index: 0,
                timestamp: BigInt(-2),
                maxNumOffsets: 1,
              },
            ],
          },
        ],
      },
    ],
    [
      "multiple offsets",
      {
        replicaId: 333,
        topics: [
          {
            name: "topic-one",
            partitions: [
              {
                index: 0,
                timestamp: BigInt(0),
                maxNumOffsets: 64,
              },
            ],
          },
        ],
      },
    ],
  ];

  test.each(cases)("%s", (_name, value) => {
    testEncodeDecodeSnapshot(
      value,
      encodeKafkaListOffsetsRequest,
      decodeKafkaListOffsetsRequest
    );
  });
});

describe("KafkaListOffsetsResponse", () => {
  type TestCase = [string, KafkaListOffsetsResponse];
  const cases: TestCase[] = [
    [
      "one offset",
      {
        topics: [
          {
            name: "topic-one",
            partitions: [
              {
                index: 0,
                errorCode: ErrorCode.None,
                oldStyleOffsets: [BigInt(16)],
              },
            ],
          },
        ],
      },
    ],
    [
      "multiple offsets",
      {
        topics: [
          {
            name: "topic-one",
            partitions: [
              {
                index: 0,
                errorCode: ErrorCode.None,
                oldStyleOffsets: [BigInt(3), BigInt(2), BigInt(1), BigInt(0)],
              },
            ],
          },
        ],
      },
    ],
    [
      "offset out of range",
      {
        topics: [
          {
            name: "topic-one",
            partitions: [
              {
                index: 0,
                errorCode: ErrorCode.OffsetOutOfRange,
                oldStyleOffsets: [],
              },
            ],
          },
        ],
      },
    ],
  ];

  test.each(cases)("%s", (_name, value) => {
    testEncodeDecodeSnapshot(
      value,
      encodeKafkaListOffsetsResponse,
      decodeKafkaListOffsetsResponse
    );
  });
});
