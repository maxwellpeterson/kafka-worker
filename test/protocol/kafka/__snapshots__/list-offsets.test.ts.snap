// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KafkaListOffsetsRequest earliest offset 1`] = `"AAABTQAAAAEACXRvcGljLW9uZQAAAAEAAAAA//////////4AAAAB"`;

exports[`KafkaListOffsetsRequest latest offset 1`] = `"AAABTQAAAAEACXRvcGljLW9uZQAAAAEAAAAA//////////8AAAAB"`;

exports[`KafkaListOffsetsRequest multiple offsets 1`] = `"AAABTQAAAAEACXRvcGljLW9uZQAAAAEAAAAAAAAAAAAAAAAAAABA"`;

exports[`KafkaListOffsetsResponse multiple offsets 1`] = `"AAAAAQAJdG9waWMtb25lAAAAAQAAAAAAAAAAAAQAAAAAAAAAAwAAAAAAAAACAAAAAAAAAAEAAAAAAAAAAA=="`;

exports[`KafkaListOffsetsResponse offset out of range 1`] = `"AAAAAQAJdG9waWMtb25lAAAAAQAAAAAAAQAAAAA="`;

exports[`KafkaListOffsetsResponse one offset 1`] = `"AAAAAQAJdG9waWMtb25lAAAAAQAAAAAAAAAAAAEAAAAAAAAAEA=="`;
